includes/REST_API/Embed_Controller.php

```diff
@@ -0,0 +1,318 @@
+		}
+
+		/**
+		 * Filters the link data TTL value.
+		 *
+		 * @param int $time Time to live (in seconds). Default is 1 day.
+		 * @param string $url The attempted URL.
+		 */
+		$cache_ttl = apply_filters( 'web_stories_embed_data_cache_ttl', DAY_IN_SECONDS, $url );
+		$cache_key = 'web_stories_embed_data_' . md5( $url );
+
+		$data = get_transient( $cache_key );
+		if ( ! empty( $data ) ) {
+			if ( 'rest_invalid_story' === $data ) {
+				return new WP_Error( 'rest_invalid_story', __( 'URL is not a story', 'web-stories' ), [ 'status' => 404 ] );
+			}
+
+			return rest_ensure_response( json_decode( $data, true ) );
+		}
+
+		$data = $this->get_data_from_post( $url );
+		if ( $data ) {
+			return rest_ensure_response( $data );
+		}
+
+		$args = [
+			'limit_response_size' => 153600, // 150 KB.
+			'timeout'             => 7, // phpcs:ignore WordPressVIPMinimum.Performance.RemoteRequestTimeout.timeout_timeout
+		];
+
+		/**
+		 * Filters the HTTP request args for link data retrieval.
+		 *
+		 * Can be used to adjust timeout and response size limit.
+		 *
+		 * @param array $args Arguments used for the HTTP request
+		 * @param string $url The attempted URL.
+		 */
+		$args = apply_filters( 'web_stories_embed_data_request_args', $args, $url );
+
+		$response = wp_safe_remote_get( $url, $args );
+
+		if ( WP_Http::OK !== wp_remote_retrieve_response_code( $response ) ) {
+			// Not saving the error response to cache since the error might be temporary.
+			return new WP_Error( 'rest_invalid_url', __( 'Invalid URL', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		$html = wp_remote_retrieve_body( $response );
+
+		if ( ! $html ) {
+			set_transient( $cache_key, 'rest_invalid_story', $cache_ttl );
+			return new WP_Error( 'rest_invalid_story', __( 'URL is not a story', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		$data = $this->get_data_from_document( $html );
+
+		if ( ! $data ) {
+			set_transient( $cache_key, 'rest_invalid_story', $cache_ttl );
+			return new WP_Error( 'rest_invalid_story', __( 'URL is not a story', 'web-stories' ), [ 'status' => 404 ] );
+		}
+
+		set_transient( $cache_key, wp_json_encode( $data ), $cache_ttl );
+
+		return rest_ensure_response( $data );
+	}
+
+	/**
+	 * Retrieves the story metadata for a given URL on the current site.
+	 *
+	 * @param string $url  The URL that should be inspected for metadata.
+	 * @return array|false Story metadata if the URL does belong to the current site. False otherwise.
+	 */
+	private function get_data_from_post( $url ) {
+		$switched_blog = false;
+
+		if ( is_multisite() ) {
+			$url_parts = wp_parse_args(
+				wp_parse_url( $url ),
+				[
+					'host' => '',
+					'path' => '/',
+				]
+			);
+
+			$qv = [
+				'domain'                 => $url_parts['host'],
+				'path'                   => '/',
+				'update_site_meta_cache' => false,
+			];
+
+			// In case of subdirectory configs, set the path.
+			if ( ! is_subdomain_install() ) {
```

Comment:
What about multi network?

Reply:
For reference, this is largely based on `get_oembed_response_data_for_url`.

If we don't catch all cases it's not a big deal, it will just result in an HTTP request.

Reply:
Okay, let's leave this for now.